src/components/modals/turbo/TurboTopUpModal.tsx

```diff
@@ -0,0 +1,233 @@
+import FiatPayment, { PaymentInformation } from './panels/fiat/FiatPayment';
+import FiatTopupComplete from './panels/fiat/FiatTopupComplete';
+
+export const errorSubmittingTransactionToTurbo =
+  'Error submitting transaction to Turbo. Please try again or contact support.';
+export const valueStringDefault = '$0 = 0 credits \u{02248} 0 GB';
+export const valueStringError = `Error: Unable to fetch credit estimate`;
+
+export type PanelStates =
+  | 'configure'
+  | 'payment'
+  | 'confirm'
+  | 'success'
+  | 'resume-eth-topup';
+type Currency = 'fiat' | 'crypto';
+
+function TurboTopUpModal({ onClose }: { onClose: () => void }) {
+  const [{ wallet, walletAddress }] = useWalletState();
+  const [currency, setCurrency] = useState<Currency>('fiat');
+
+  const [panelState, setPanelState] = useState<PanelStates>('configure');
+
+  const [paymentIntent, setPaymentIntent] = useState<PaymentIntent>();
+  const [paymentAmount, setPaymentAmount] = useState<number>();
+  const [promoCode, setPromoCode] = useState<string>();
+  const [, setCryptoTopupValue] = useState<number>();
+  const [paymentInformation, setPaymentInformation] =
+    useState<PaymentInformation>();
+  const [paymentIntentError, setPaymentIntentError] = useState<string>();
+
+  const [valueString, setValueString] = useState<string>(valueStringDefault);
+  const [topupValue, setTopupValue] = useState<number>();
+
+  const [, setPaymentIntentResult] = useState<PaymentIntentResult>();
+
+  useEffect(() => {
+    async function updatePaymentIntent() {
+      if (currency === 'fiat') {
+        if (!walletAddress || !wallet?.tokenType || !topupValue) {
+          return;
+        }
+        try {
+          const paymentIntent = await getPaymentIntent(
+            walletAddress.toString(),
+            USD(topupValue).amount,
+            wallet.tokenType,
+          );
+          if (paymentIntent.paymentSession) {
+            setPaymentIntent(paymentIntent.paymentSession);
+            setPaymentAmount(USD(topupValue).amount);
+          }
+        } catch (e: unknown) {
+          setPaymentIntentError((e as Error).message);
+          return;
+        }
+      } else {
+        setCryptoTopupValue(topupValue!);
+      }
+    }
+    updatePaymentIntent();
+  }, [walletAddress, topupValue, wallet]);
+
+  useEffect(() => {
+    // Prevent user from leaving the page when payment is in progress, requires confirmation
+    const handleBeforeUnload = (e: BeforeUnloadEvent) => {
+      if (panelState === 'payment' || panelState === 'confirm') {
+        e.preventDefault();
+      }
+    };
+    window.addEventListener('beforeunload', handleBeforeUnload);
+    return () => {
+      window.removeEventListener('beforeunload', handleBeforeUnload);
+    };
+  }, [panelState]);
+
+  function handleClose() {
+    if (panelState === 'payment' || panelState === 'confirm') {
+      if (window.confirm('Are you sure you want to close this payment?')) {
```

Comment:
Using this might look a bit jarring coming from close button state in-app vs. custom in-app modal that is styled. Something to look at another time as tech debt to polish this a bit.